gh-81074: Allow non-ASCII addr_spec in email.headerregistry.Address

medmunds opened the pull request:
The email.headerregistry.Address constructor raised an error if addr_spec contained a non-ASCII character. (But it fully supports non-ASCII in the separate username and domain args.) This change removes the error for a non-ASCII addr_spec.

This PR implements @bitdancer's suggested fix from https://github.com/python/cpython/issues/81074#issuecomment-1093823543:

> … But with RFC6532 support, it should be valid to have a local part that has non-ascii in an Address, and the error, as I noted above, should be raised only at serialization time and when we don't have an original source string. So that raise should be modified to explicitly ignore the NonASCIILocalPartDefect.

(The other bugs discussed in that comment are reported separately as #83938 and #122476.)

<!-- gh-issue-number: gh-81074 -->
* Issue: gh-81074
<!-- /gh-issue-number -->

Fixes gh-81074

medmunds committed:
Allow non-ASCII addr_spec in email.headerregistry.Address

The email.headerregistry.Address constructor raised an error if
addr_spec contained a non-ASCII character. (But it fully supports
non-ASCII in the separate username and domain args.) This change
removes the error for a non-ASCII addr_spec.

medmunds committed:
Allow non-ASCII addr_spec in email.headerregistry.Address

The email.headerregistry.Address constructor raised an error if
addr_spec contained a non-ASCII character. (But it fully supports
non-ASCII in the separate username and domain args.) This change
removes the error for a non-ASCII addr_spec.

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

medmunds committed:
Merge branch 'main' into fix-issue-81074

picnixz committed:
Merge branch 'main' into fix-issue-81074

bitdancer commented on Lib/email/headerregistry.py:
```diff
@@ -45,8 +45,12 @@ def __init__(self, display_name='', username='', domain='', addr_spec=None):
+                if not isinstance(defect, errors.NonASCIILocalPartDefect)
+            ]
+            if relevant_defects:
+                raise relevant_defects[0]
```
I started to write that this should have a comment because the context was lacking, but when I wrote it I realized what this really means is that NonASCIILocalPartDefect is not a true defect.  Especially when you consider that the comments say it can only happen when parsing unicode, not when parsing bytes.  So instead of doing this, now that we're doing the serialization check (which is where it becomes a true defect) we should just remove the check for this defect in _header_value_parser.

We can't remove it from errors.py for backward compatibility reasons, but it should be marked as deprecated.

medmunds replied:
I had been wondering why there wasn't a corresponding NonASCIIDomainDefect. I've removed the parser check and restored this code in Address.\_\_init\_\_() to its original form.

medmunds committed:
Merge branch 'main' into fix-issue-81074

medmunds committed:
Merge remote-tracking branch 'origin/fix-issue-81074' into fix-issue-81074

medmunds committed:
Stop using NonASCIILocalPartDefect in email._header_value_parser

When parsing email messages from Unicode strings (but not bytes),
get_local_part() recorded a NonASCIILocalPartDefect for non-ASCII
characters. RFC 5322 permits such addresses.

This change:
- removes the parse-time detection for a non-ASCII local-part
  (and a related test)
- adds tests for passing a non-ASCII addr_spec to
  email.headerregistry.Address.__init__()
- marks the (undocumented) email.errors.NonASCIILocalPartDefect
  as unused and deprecated

 This affected parsing email messages from Unicode strings
(but not from bytes), and also prevented
